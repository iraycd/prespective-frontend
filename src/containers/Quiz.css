.medium {
  color: orange;
}

.rating {
  width: 16px;
  height: 16px;
  background-color: #fff;
  border: 3px solid #d2e1ee;
  border-radius: 100%;
  -webkit-transition: border-width 0.25s;
  transition: border-width 0.25s;
  display: inline-block;
}

.rating.minus-three {
  border-color: #4c02a1;
  background-color: #4c02a1;
  width: 32px;
  height: 32px;
}
.rating.minus-two {
  border-color: #7700ff;
  background-color: #7700ff;
  width: 24px;
  height: 24px;
}
.rating.minus-one {
  border-color: #ab62ff;
  background-color: #ab62ff;
  width: 16px;
  height: 16px;
}
.rating.neutral {
  border-color: #d2e1ee;
  background-color: #d2e1ee;
  width: 12px;
  height: 12px;
}
.rating.plus-three {
  border-color: #b33300;
  background-color: #b33300;
  width: 32px;
  height: 32px;
}
.rating.plus-two {
  border-color: #ff5e1e;
  background-color: #ff5e1e;
  width: 24px;
  height: 24px;
}
.rating.plus-one {
  border-color: #ff9369;
  background-color: #ff9369;
  width: 16px;
  height: 16px;
}
.rating.empty {
  background-color: #fff;
}
