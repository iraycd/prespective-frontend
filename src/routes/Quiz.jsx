import React from 'react';

import Github from 'containers/GitHub';

import { Box, Container, Heading, Screen, Text } from 'styled-minimal';

const Quiz = () => (
  <Screen key="Quiz" data-testid="QuizWrapper">
    <Container verticalPadding>
      <Box mb={4}>
        <Heading as="h5">Discover Your Perspective</Heading>
        <Text fontSize={1}>
          Complete the 7 min test and get a detailed report of your lenses on the world.
        </Text>
      </Box>
      <Github />
    </Container>
  </Screen>
);

export default Quiz;
