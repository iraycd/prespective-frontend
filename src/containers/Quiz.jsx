import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';
import treeChanges from 'tree-changes';
import ReactRating from 'react-rating';
import { appColor } from 'modules/theme';

import { getQuestions, showAlert, submitQuiz } from 'actions/index';
import { STATUS } from 'constants/index';

import { Heading, Button, Input, Label, Text, theme, utils } from 'styled-minimal';
import Loader from 'components/Loader';
import './Quiz.css';

const { spacer } = utils;
const { grays } = theme;

const QuizGrid = styled.ul`
  display: grid;
  grid-auto-flow: row;
  grid-template-columns: 100%;
  list-style: none;
  padding: 0;
  > li {
    display: flex;
  }
`;

const EmailGrid = styled.div`
  width: 400px;
  margin: 0px auto;
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-direction: column;
  > {
    margin: 10px;
  }
`;

const Item = styled.span`
  align-items: center;
  border: solid 0.1rem ${appColor};
  border-radius: 0.4rem;
  overflow: hidden;
  padding: ${spacer(3)};
  text-align: center;
  width: 100%;
  p {
    color: #000;
  }

  img {
    height: 8rem;
    margin-bottom: ${spacer(2)};
  }
`;

const ItemHeader = styled.div`
  margin-bottom: ${spacer(3)};

  small {
    color: ${grays.gray60};
  }
`;

const Rating = styled(ReactRating)`
  margin: 30px auto;
  width: 50%;
  display: flex !important;
  justify-content: space-around;
  align-items: center;
`;

const SubmitButton = styled(Button)`
  border-radius: 0.3rem;
  background-color: #3d59fa;
  border-color: #3d59fa;
  border-radius: 0.6rem;
  color: #fff;
  :hover {
    color: #fff;
    background-color: #1839f9;
    border-color: #0b2ef9;
  }
`;

export class Quiz extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      answerMap: {},
    };
  }

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    history: PropTypes.object,
    quiz: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(getQuestions());
  }

  componentWillReceiveProps(nextProps) {
    const { dispatch, history } = this.props;
    const { changedTo } = treeChanges(this.props, nextProps);
    if (nextProps.quiz.resultId !== '') {
      history.push(`/result/${nextProps.quiz.resultId}`);
    }
    if (changedTo('quiz', STATUS.ERROR)) {
      dispatch(showAlert(nextProps.quiz.questions.message, { variant: 'danger' }));
    }
  }

  setEmail = email => {
    this.setState({ email });
  };

  setAnswer = (questionId, answer) => {
    let { answerMap } = this.state;
    answerMap = { ...answerMap, [questionId]: answer };
    this.setState({
      answerMap,
    });
  };

  submitAnswers = () => {
    const { dispatch, quiz } = this.props;
    const { answerMap, email } = this.state;
    const answerList = [];
    if (quiz.questions.length === Object.keys(answerMap).length) {
      Object.entries(answerMap).forEach(([key, value]) => {
        answerList.push({
          answer: value,
          question: key,
        });
      });
      dispatch(submitQuiz(email, answerList));
    } else {
      dispatch(showAlert('All question are not answered', { variant: 'danger' }));
    }
  };

  render() {
    const { quiz } = this.props;
    const { answerMap } = this.state;
    const data = quiz.questions || [];
    let output;

    if (quiz.status === STATUS.READY) {
      if (data.length) {
        output = (
          <QuizGrid>
            {quiz.questions.map(d => (
              <li key={d._id}>
                <Item>
                  {/* <Image src={d.owner.avatar_url} alt={d.owner.login} /> */}
                  <ItemHeader>
                    <Heading as="h5" lineHeight={1}>
                      {d.question}
                    </Heading>
                    <Rating
                      initialRating={answerMap[d._id]}
                      start={0}
                      stop={7}
                      onClick={rate => this.setAnswer(d._id, rate)}
                      emptySymbol={[
                        'rating empty minus-three',
                        'rating empty minus-two',
                        'rating empty minus-one',
                        'rating empty neutral',
                        'rating empty plus-one',
                        'rating empty plus-two',
                        'rating empty plus-three',
                      ]}
                      fullSymbol={[
                        'rating minus-three',
                        'rating minus-two',
                        'rating minus-one',
                        'rating neutral',
                        'rating plus-one',
                        'rating plus-two',
                        'rating plus-three',
                      ]}
                    />
                  </ItemHeader>
                </Item>
              </li>
            ))}
          </QuizGrid>
        );
      } else {
        output = <h3>Nothing found</h3>;
      }
    } else {
      output = <Loader block />;
    }
    return (
      <div key="Quiz" data-testid="QuizWrapper">
        {output}

        <EmailGrid>
          <Label>Your Email</Label>
          <Input placeholder="you@example.com" onChange={e => this.setEmail(e.target.value)} />
          <Text>Let's see what your result looks like.</Text>
          <SubmitButton onClick={this.submitAnswers}>Save and Continue</SubmitButton>
        </EmailGrid>
      </div>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return { quiz: state.quiz };
}

export default connect(mapStateToProps)(Quiz);
