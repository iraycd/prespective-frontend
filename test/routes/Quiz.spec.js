import React from 'react';

import Quiz from 'routes/Quiz';

function setup() {
  const props = {
    dispatch: () => {},
    location: {},
  };

  return shallow(<Quiz {...props} />);
}

describe('Quiz', () => {
  const wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
