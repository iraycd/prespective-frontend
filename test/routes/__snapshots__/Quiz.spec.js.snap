// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz should render properly 1`] = `
<Screen
  data-testid="QuizWrapper"
  key="Quiz"
  minHeight="100vh"
>
  <Container
    verticalPadding={true}
  >
    <Box
      mb={4}
    >
      <Heading
        as="h5"
      >
        Discover Your Perspective
      </Heading>
      <Text
        as="span"
        fontSize={1}
      >
        Complete the 7 min test and get a detailed report of your lenses on the world.
      </Text>
    </Box>
    <Connect(Quiz) />
  </Container>
</Screen>
`;
